system_metrics/system_metrics_write_closer_test.go

```diff
@@ -0,0 +1,302 @@
+
+		testString := []byte("large test string")
+		n, err := w.Write(testString)
+		require.NoError(t, err)
+		assert.Equal(t, len(testString), n)
+		require.NotNil(t, mc.GetData())
+		assert.Equal(t, testString, mc.GetData().Data)
+		raw.mu.Lock()
+		assert.Equal(t, []byte{}, raw.buffer)
+		raw.mu.Unlock()
+		assert.NoError(t, w.Close())
+	})
+	t.Run("ExistingDataInBuffer", func(t *testing.T) {
+		mc := &mockClient{}
+		s := &SystemMetricsClient{client: mc}
+
+		w, err := s.NewSystemMetricsWriteCloser(ctx, dataOpts, WriteCloserOptions{
+			NoTimedFlush: true,
+		})
+		require.NoError(t, err)
+		raw, ok := w.(*systemMetricsWriteCloser)
+		require.True(t, ok)
+
+		testString1 := []byte("first test string")
+		testString2 := []byte("second test string")
+		_, err = w.Write(testString1)
+		require.NoError(t, err)
+		n, err := w.Write(testString2)
+		require.NoError(t, err)
+		assert.Equal(t, len(testString2), n)
+		raw.mu.Lock()
+		assert.Equal(t, append(testString1, testString2...), raw.buffer)
+		raw.mu.Unlock()
+		assert.NoError(t, w.Close())
+	})
+	t.Run("AlreadyClosed", func(t *testing.T) {
+		mc := &mockClient{}
+		s := &SystemMetricsClient{client: mc}
+
+		w, err := s.NewSystemMetricsWriteCloser(ctx, dataOpts, WriteCloserOptions{
+			NoTimedFlush: true,
+		})
+		require.NoError(t, err)
+		require.NoError(t, w.Close())
+
+		n, err := w.Write([]byte("small test string"))
+		assert.Error(t, err)
+		assert.Equal(t, 0, n)
+	})
+	t.Run("NoData", func(t *testing.T) {
+		mc := &mockClient{}
+		s := &SystemMetricsClient{client: mc}
+
+		w, err := s.NewSystemMetricsWriteCloser(ctx, dataOpts, WriteCloserOptions{
+			NoTimedFlush: true,
+		})
+		require.NoError(t, err)
+
+		n, err := w.Write([]byte{})
+		require.Error(t, err)
+		assert.Equal(t, 0, n)
+		assert.NoError(t, w.Close())
+	})
+	t.Run("RPCError", func(t *testing.T) {
+		mc := &mockClient{addErr: true}
+		s := &SystemMetricsClient{client: mc}
+
+		w, err := s.NewSystemMetricsWriteCloser(ctx, dataOpts, WriteCloserOptions{
+			NoTimedFlush:  true,
+			MaxBufferSize: 1,
+		})
+		require.NoError(t, err)
+
+		n, err := w.Write([]byte("small test string"))
+		assert.Error(t, err)
+		assert.Equal(t, 0, n)
+	})
+}
+
+func TestSystemMetricsWriteCloserTimedFlush(t *testing.T) {
+	ctx := context.Background()
+	dataOpts := MetricDataOptions{
+		Id:         "ID",
+		MetricType: "Test",
+		Format:     DataFormatFTDC,
+	}
+	t.Run("ValidOutput", func(t *testing.T) {
+		mc := &mockClient{}
+		s := &SystemMetricsClient{client: mc}
+
+		w, err := s.NewSystemMetricsWriteCloser(ctx, dataOpts, WriteCloserOptions{
+			FlushInterval: time.Second,
+		})
+		require.NoError(t, err)
+		raw, ok := w.(*systemMetricsWriteCloser)
+		require.True(t, ok)
+
+		testString := []byte("small test string")
+		_, err = w.Write(testString)
+		require.NoError(t, err)
+		assert.Nil(t, mc.GetData())
+		time.Sleep(2 * time.Second)
+		require.NotNil(t, mc.GetData())
+		assert.Equal(t, testString, mc.GetData().Data)
+		raw.mu.Lock()
```

Comment:
I don't think its worth it to write such a function since in some cases I am intentionally manipulating data in the underlying struct (so I need to access the data directly rather than read a copy). The other thing I could do is write safe getters for the fields I check but that seems excessive for a test.